### ISSUE TITLE
Unhandled promise rejection because of createConnection initial connection promise rejection

### ISSUE THREAD

**Issue by bbhopesh** [Automattic/mongoose#8706]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
If there is an initial connection error in createConnection, it generates an Unhandled promise rejection. There are two workarounds but they are quite hacky.

**If the current behavior is a bug, please provide the steps to reproduce.**

Get gist [here](https://gist.github.com/bbhopesh/5d443784c527a30ff829140c7911a299)
MongoDB shouldn't be running to generate initial connection error.

**What is the expected behavior?**
You will see two test cases in the gist, the second one is commented and first one is active.   

**Test Case I:**  In this test case, `createConnection` is not able to establish initial connection to database and you get Unhandled rejected promise warning.  
**Test Case II:** In this test case, error and then handlers are provided and you don't get unhandled rejected promise warning anymore but now you have to write all of code inside callback or an async function which is very ugly.

There should be a simpler way to deal with initial connection errors while using `createConnection`. `mongoose.connect` doesn't have same problem because even if you provide `then` and `catch` while connecting, you can get handle to connection outside callback using `mongoose.connection`

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node.js: 12.13.0
Mongoose: 5.8.12
MongoDB: 4.0.12

**Comment by vkarpov15** [comment]
Call me crazy, but why don't you just use [`Promise#catch()` to catch the promise rejection](https://masteringjs.io/tutorials/fundamentals/catch)?

```javascript
const conn = mongoose.createConnection('mongodb://localhost:27017');
conn.catch(err => console.log(err));
```

**Comment by bbhopesh** [comment]
Because by the time you try to call `catch`, it is already deleted from the connection, the `catch` and `then` functions are deleted after initial connection promise resolves or rejects. See [here](https://github.com/Automattic/mongoose/blob/055b75da23d1a92bb31de19679f4f8c0001cd078/lib/connection.js#L809)

**Comment by bbhopesh** [comment]
Hi @vkarpov15 I don't mean to rush you but please take a look at this whenever you get some time.

**Comment by vkarpov15** [comment]
@bbhopesh I check github emails in reverse order, so bumping puts you at the end of the queue.

How are you making it so that you're calling `.catch()` after the connection promise is already rejected? Can you provide a code sample that demonstrates that?

Also, can you please clarify why `mongoose.createConnection('mongodb://localhost:27017').catch(err => { ... })` doesn't work for you? It shouldn't be possible for `createConnection()` to reject before that `.catch()` is called.

**Comment by bbhopesh** [comment]
Shutdown mongodb and run these two lines and you will be reproduce.
```
const conn = mongoose.createConnection('mongodb://localhost:27017');
conn.catch(err => console.log(err)); // by time this line executes, promise is already rejected and there is no catch function.
```

`mongoose.createConnection('mongodb://localhost:27017').catch(err => { ... })` works as far as error handling goes but now because of `.catch(...)` , the returned object is no longer a connection. I don't have a reference to connection object. There is a hack to get it from `mongoose.connections` array but that is not reliable.

**Comment by bbhopesh** [comment]
you added the tag that it needs repro script but I provided the repro script/steps above. Are you waiting on me for something?

**Comment by AbdelrahmanHafez** [comment]
@bbhopesh 

I can't reproduce this, running:
```js
const mongoose = require('mongoose');

const conn = mongoose.createConnection('mongodb://localhost:27017');
conn.catch(err => console.log(err));
```

logs:
```
(node:23204) DeprecationWarning: current URL string parser is deprecated, and will be removed in a future version. To use the new parser, pass option { useNewUrlParser: true } to MongoClient.connect.
(node:23204) DeprecationWarning: current Server Discovery and Monitoring engine is deprecated, and will be removed in a future version. To use the new Server Discover and Monitoring engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.
MongoNetworkError: failed to connect to server [localhost:27017] on first connect [Error: connect ECONNREFUSED 127.0.0.1:27017
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1137:16) {
  name: 'MongoNetworkError',
  [Symbol(mongoErrorContextSymbol)]: {}
}]
    at Pool.<anonymous> (\node_modules\mongoose\node_modules\mongodb\lib\core\topologies\server.js:438:11)
    at Pool.emit (events.js:311:20)
    at \node_modules\mongoose\node_modules\mongodb\lib\core\connection\pool.js:561:14
    at \node_modules\mongoose\node_modules\mongodb\lib\core\connection\pool.js:994:11
    at \node_modules\mongoose\node_modules\mongodb\lib\core\connection\connect.js:31:7
    at callback (\node_modules\mongoose\node_modules\mongodb\lib\core\connection\connect.js:264:5)
    at Socket.<anonymous> (\node_modules\mongoose\node_modules\mongodb\lib\core\connection\connect.js:294:7)
    at Object.onceWrapper (events.js:418:26)
    at Socket.emit (events.js:311:20)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  name: 'MongoNetworkError',
  [Symbol(mongoErrorContextSymbol)]: {}
}
```

**Comment by bbhopesh** [comment]
It would on the timing of initial promise getting rejected. `catch` is deleted from connection after initial connection is rejected. Try adding a delay before calling `conn.catch` and you will be able to reproduce.

**Comment by AbdelrahmanHafez** [comment]
I am not sure I understand, why would I add a delay before adding an error handler on the connection?

Could you provide some script that would demonstrate the issue with the delay?

**Comment by lineus** [comment]
@bbhopesh 

> Test Case I: In this test case, createConnection is not able to establish initial connection to database and you get Unhandled rejected promise warning.

You never attach a rejection handler to the first test case in your gist, so it's always going to be an `UnhandledPromiseRejectionWarning`. Registering an error event handler isn't the same as providing a handler for the promise rejection.


Here's an example that shows a delay in attaching the rejection handler. It appears you have until 30 seconds have elapsed to attach the rejection handler to the connection and have it work as expected ( at least for ECONNREFUSED ).

### 8706.js
```js
#!/usr/bin/env node
'use strict';

const assert = require('assert');
const mongoose = require('mongoose');
const DB = '8706';
const URI = `mongodb://localhost:27018/${DB}`; // no mongodb here
const OPTS = { useNewUrlParser: true, useUnifiedTopology: true };

// if this value is 30 seconds or higher
// you get an UnhandledPromiseRejectionWarning
const WAIT = 1000 * 30;

async function run() {
  const conn = mongoose.createConnection(URI, OPTS);
  setTimeout(() => {
    conn.catch(err => {
      assert.ok(/ECONNREFUSED/.test(err.message));
    });
  }, WAIT);
}

run();
```
### Output with 30 second WAIT
```
$ ./8706.js
(node:57347) UnhandledPromiseRejectionWarning: MongooseServerSelectionError: connect ECONNREFUSED 127.0.0.1:27018
    at new MongooseServerSelectionError (/Users/lineus/dev/node/mongoose/node_modules/mongoose/lib/error/serverSelection.js:22:11)
    at NativeConnection.Connection.openUri (/Users/lineus/dev/node/mongoose/node_modules/mongoose/lib/connection.js:823:32)
    at Mongoose.connect (/Users/lineus/dev/node/mongoose/node_modules/mongoose/lib/index.js:333:15)
    at run (/Users/lineus/dev/node/mongoose/issues/8706.js:15:25)
    at Object.<anonymous> (/Users/lineus/dev/node/mongoose/issues/8706.js:24:1)
    at Module._compile (internal/modules/cjs/loader.js:774:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:785:10)
    at Module.load (internal/modules/cjs/loader.js:641:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:837:10)
    at internal/main/run_main_module.js:17:11
(node:57347) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:57347) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
All Assertions Pass.
(node:57347) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 1)
$
```

### Output with 25 second WAIT:
```
$ ./8706.js
All Assertions Pass.
$ 
```

**Comment by bbhopesh** [comment]
@AbdelrahmanHafez You won't add a delay in real application. I meant it will demonstrate the race condition. 

@lineus The time window varies from machine to machine. I don't have any time window at all on my machine. Even if its next line of code without any delay, I get promise rejection and also "can't call undefined" because by the time next line is called, the promise is rejected and `catch` function is deleted from connection. Only way for me to add handler is to do `createConnection().catch(...)` but then I loose handle to connection.

**Comment by lineus** [comment]
@bbhopesh Regarding the timing, the time window shouldn't vary from machine to machine. It's dependent on whether or not you are using the new unified topology ( *and if so, the value of serverSelectionTimeoutMS which defaults to 30 seconds* ). If you're not using the new unified topology, then the initial connection attempt fails immediately upon execution.
 
> catch function is deleted from connection

This is the part no one else has been able to reproduce.

### 8706_3.js
```js
#!/usr/bin/env node --no-deprecation
'use strict';

const assert = require('assert');
const mongoose = require('mongoose');
let called = 0;

assert.strictEqual(mongoose.version, '5.8.12');

const conn = mongoose.createConnection('mongodb://localhost:27018'); // no mongo here
addHandler(conn);
assert.strictEqual(typeof conn.catch, 'function');
conn.catch(eHandler);


function addHandler(c) {
  c.on('error', eHandler);
}

function eHandler(e) {
  if (/ECONNREFUSED/.test(e.message)) {
    called++;
  }
}

process.on('exit', () => {
  assert.strictEqual(called, 2);
  console.log('All Assertions Pass.');
});
```
### Output:
```
$ ./8706_3.js
All Assertions Pass.
$
```

**Comment by vkarpov15** [comment]
I opened up a separate issue #8810 to track a potential fix. @bbhopesh is right in that if you wait for too long to add a `.catch()` to a connection, Mongoose will delete the `.catch()` method. That was an unfortunate hack from https://github.com/Automattic/mongoose/commit/343b7a79651b4fe75e5e9b03d92659a76de323b3 to support #5404 without being backwards breaking.

**Comment by bbhopesh** [comment]
Thanks @vkarpov15 .
@lineus I wasn't using unified topology but when you don't use unified topology "immediately" could vary from machine to machine.
